addons/account/wizard/account_merge_wizard.py

```diff
@@ -0,0 +1,310 @@
+
+            wizard_lines_vals_list = []
+            sequence = 0
+            for grouping_key, group_accounts in accounts.grouped(wizard._get_grouping_key).items():
+                grouping_key_str = str((i, *grouping_key))  # The grouping key identifies both the wizard and the account properties
+                wizard_lines_vals_list.append({
+                    'display_type': 'line_section',
+                    'info': wizard._get_group_name(group_accounts[0]),
+                    'grouping_key': grouping_key_str,
+                    'sequence': (sequence := sequence + 1),
+                })
+                for account in group_accounts:
+                    wizard_lines_vals_list.append({
+                        'display_type': 'account',
+                        'account_id': account.id,
+                        'grouping_key': grouping_key_str,
+                        'is_selected': True,
+                        'sequence': (sequence := sequence + 1),
+                    })
+
+            wizard.wizard_line_ids = [
+                Command.unlink(wizard_line.id)
+                for wizard_line in wizard.wizard_line_ids
+            ] + [
+                Command.create(vals)
+                for vals in wizard_lines_vals_list
+            ]
+
+    @api.depends('wizard_line_ids.is_selected', 'wizard_line_ids.info')
+    def _compute_disable_merge_button(self):
+        for wizard in self:
+            wizard_lines_to_merge = wizard.wizard_line_ids.filtered(lambda l: l.display_type == 'account' and l.is_selected and not l.info)
+            wizard.disable_merge_button = all(
+                len(wizard_line_group) < 2
+                for wizard_line_group in wizard_lines_to_merge.grouped('grouping_key').values()
+            )
+
+    def action_merge(self):
+        """ Merge each group of accounts in `self.wizard_line_ids`. """
+        self._check_access_rights(self.account_ids)
+
+        for wizard in self:
+            wizard_lines_selected = wizard.wizard_line_ids.filtered(lambda l: l.display_type == 'account' and l.is_selected and not l.info)
+            for wizard_lines_group in wizard_lines_selected.grouped('grouping_key').values():
+                if len(wizard_lines_group) > 1:
+                    self._action_merge(wizard_lines_group.account_id)
+
+        return {
+            'type': 'ir.actions.client',
+            'tag': 'display_notification',
+            'params': {
+                'type': 'success',
+                'sticky': False,
+                'message': _("Accounts successfully merged!"),
+                'next': {'type': 'ir.actions.act_window_close'},
+            }
+        }
+
+    @api.model
+    def _check_access_rights(self, accounts):
+        accounts.check_access_rights('write')
+        accounts.check_access_rule('write')
+        if forbidden_companies := (accounts.sudo().company_ids - self.env.user.company_ids):
```

Comment:
Isn't this part of `check_access_rule`?

Reply:
I don't have the impression it is?
The access rule is that you can access the account as long as long as there is overlap between the active company and the account's `company_ids`. 
Here, we additionally need to have access to *all* of the account's `company_ids`.